src/awkward/contents/numpyarray.py

```diff
@@ -340,59 +340,34 @@ def _offsets_and_flattened(self, axis, depth):
             )
 
     def _mergeable_next(self, other, mergebool):
-        if isinstance(
-            other,
-            (
-                ak.contents.IndexedArray,
-                ak.contents.IndexedOptionArray,
-                ak.contents.ByteMaskedArray,
-                ak.contents.BitMaskedArray,
-                ak.contents.UnmaskedArray,
-            ),
-        ):
-            return self._mergeable(other._content, mergebool)
+
+        if len(self.shape) > 1:
+            return self._to_regular_primitive()._mergeable(other, mergebool)
 
         elif isinstance(other, ak.contents.NumpyArray):
             if self._data.ndim != other._data.ndim:
                 return False
 
-            matching_dtype = self._data.dtype == other._data.dtype
-
-            if (
-                not mergebool
-                and not matching_dtype
-                and (
-                    self._data.dtype.type is np.bool_
-                    or other._data.dtype.type is np.bool_
-                )
-            ):
-                return False
+            # Obvious fast-path
+            if self.dtype == other.dtype:
+                return True
 
-            if not matching_dtype and np.datetime64 in (
-                self._data.dtype.type,
-                other._data.dtype.type,
+            # Special-case booleans i.e. {bool, number}
+            elif mergebool and (
+                np.issubdtype(self.dtype, np.bool_)
+                and np.issubdtype(other.dtype, np.number)
+                or np.issubdtype(self.dtype, np.number)
+                and np.issubdtype(other.dtype, np.bool_)
             ):
-                return False
+                return True
 
-            if not matching_dtype and np.timedelta64 in (
-                self._data.dtype.type,
-                other._data.dtype.type,
-            ):
-                return False
-
-            if (
-                len(self._data.shape) > 1
-                and self._data.shape[1:] != other._data.shape[1:]
-            ):
-                return False
-
-            return True
-
-        # If we have >1 dimension, promote ourselves to `RegularArray` and attempt to merge.
-        elif isinstance(other, ak.contents.RegularArray) and self.purelist_depth > 1:
-            as_regular_array = self.to_RegularArray()
-            assert isinstance(as_regular_array, ak.contents.RegularArray)
-            return as_regular_array._content._mergeable(other._content, mergebool)
+            # Default merging (can we cast one to the other)
+            else:
+                return self.backend.nplike.can_cast(
+                    self.dtype, other.dtype, casting="same_kind"
+                ) or self.backend.nplike.can_cast(
+                    other.dtype, self.dtype, casting="same_kind"
+                )
```

Comment:
It's specifically this part of the code (line 350 to here) that was tuned to give the right merging behavior for datetime and timedelta types. It was originally wrong; I fixed it here:

https://github.com/scikit-hep/awkward/pull/1183/commits/0734ab26a47e65548f84f750225fa0eb4adb435f#diff-cdc2b042962b124ea98cc8bacd6aea030a50dc73f6502f9686475554c8913320

("Wrong" was mixing datetimes or timedeltas of different units without conversion.)

As I said in today's meeting, I'd like to continue being stricter than NumPy, not allowing datetimes or timedeltas to be mixed with non-datetimes or non-timedeltas. When the Array API Consortium eventually gets around to defining time-unit behavior, I suspect that they'll come to the same conclusion: promoting non-time data to act as time-data means _guessing_ a unit, which usually isn't safe. If I'm right and they _do_ pick this behavior, we won't have to change it. (Since it's a behavior without an API marker, like a function argument, it would be hard to set up a deprecation cycle.)

So let's revert the use of `nplike.can_cast`, at least for time units or any conversions that are already in violation with Array API decisions. We can diverge from current-NumPy if we think doing so means making smaller or no changes later, when NumPy is compliant with Array API.

Reply:
I'm surprised, though, that the above didn't break any tests. I remember there being some tests that were sensitive to the time-unit fix, but I can't find them now.